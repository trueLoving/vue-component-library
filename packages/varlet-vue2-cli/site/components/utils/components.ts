import Vue, { ComponentOptions } from 'vue'
import { CombinedVueInstance } from 'vue/types/vue'

export function pickProps<T, U extends keyof T>(props: T, propsKey: U): T[U]
export function pickProps<T, U extends keyof T>(props: T, propsKey: U[]): Pick<T, U>
export function pickProps(props: any, propsKey: any): any {
  return Array.isArray(propsKey)
    ? propsKey.reduce((pickedProps: any, key) => {
        pickedProps[key] = props[key]
        return pickedProps
      }, {})
    : props[propsKey]
}

export interface MountComponentApi {
  instance: CombinedVueInstance<any, any, any, any, any>
  unmount(): void
}

export function mountComponent(component: ComponentOptions<Vue>, container = 'body', options = {}): MountComponentApi {
  const instance = new (Vue.extend(component))(options)
  const el = instance.$mount().$el
  const wrapper = document.querySelector(container)!
  wrapper.appendChild(el)

  return {
    instance,
    unmount() {
      instance.$destroy()
      wrapper.removeChild(el)
    },
  }
}
