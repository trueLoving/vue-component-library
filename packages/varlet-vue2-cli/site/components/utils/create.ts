import Vue, { ComponentOptions, RenderContext, VueConstructor } from 'vue'
import { SlotsMixin } from './mixins/slots'
import { ListenersMixin } from './mixins/listeners'

export type ComponentOptionsPlugin = ComponentOptions<Vue> & { install?: (app: VueConstructor) => void }

export function nextTick() {
  return new Promise(Vue.nextTick)
}

export function mergeSlots(context: RenderContext) {
  const scopedSlots = context.scopedSlots ?? context.data.scopedSlots ?? {}
  const slots = context.slots()

  Object.keys(slots).forEach((key) => {
    if (!scopedSlots[key]) {
      scopedSlots[key] = () => slots[key]
    }
  })

  return scopedSlots
}

export function defineComponent(sfc: ComponentOptionsPlugin & { name: string }) {
  if (sfc.mixins) {
    sfc.mixins.push(SlotsMixin)
    sfc.mixins.push(ListenersMixin)
  } else {
    sfc.mixins = [SlotsMixin, ListenersMixin]
  }

  return sfc
}
