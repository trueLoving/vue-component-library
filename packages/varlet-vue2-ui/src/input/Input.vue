<template>
  <div class="var-input var--box" :class="[disabled ? 'var-input--disabled' : null]" @click="handleClick">
    <div
      class="var-input__controller"
      :class="[
        isFocus ? 'var-input--focus' : null,
        errorMessage ? 'var-input--error' : null,
        formDisabled || disabled ? 'var-input--disabled' : null,
      ]"
      :style="{
        color: !errorMessage ? (isFocus ? focusColor : blurColor) : undefined,
      }"
    >
      <div class="var-input__icon" :class="[!hint ? 'var-input--non-hint' : null]">
        <slot name="prepend-icon" />
      </div>

      <div class="var-input__wrap" :class="[!hint ? 'var-input--non-hint' : null]">
        <input class="var-input__autocomplete" v-if="type === 'password'" />
        <textarea
          class="var-input__input var-input--textarea"
          ref="el"
          autocomplete="new-password"
          :id="id"
          :disabled="formDisabled || disabled || formReadonly || readonly"
          :type="type"
          :value="value"
          :maxlength="maxlength"
          :rows="rows"
          :class="[
            formDisabled || disabled ? 'var-input--disabled' : null,
            errorMessage ? 'var-input--caret-error' : null,
          ]"
          :style="{
            color: textColor,
            caretColor: !errorMessage ? focusColor : null,
            resize: resize ? 'vertical' : 'none',
          }"
          @focus="handleFocus"
          @blur="handleBlur"
          @input="handleInput"
          @change="handleChange"
          v-if="textarea"
        >
        </textarea>
        <input
          class="var-input__input"
          ref="el"
          autocomplete="new-password"
          :id="id"
          :disabled="formDisabled || disabled || formReadonly || readonly"
          :type="type"
          :value="value"
          :maxlength="maxlength"
          :class="[
            formDisabled || disabled ? 'var-input--disabled' : null,
            errorMessage ? 'var-input--caret-error' : null,
          ]"
          :style="{
            color: textColor,
            caretColor: !errorMessage ? focusColor : null,
          }"
          @focus="handleFocus"
          @blur="handleBlur"
          @input="handleInput"
          @change="handleChange"
          v-else
        />
        <label
          class="var--ellipsis"
          :class="[
            isFocus ? 'var-input--focus' : null,
            errorMessage ? 'var-input--error' : null,
            textarea ? 'var-input__textarea-placeholder' : 'var-input__placeholder',
            computePlaceholderState(),
            !hint ? 'var-input--placeholder-non-hint' : null,
          ]"
          :style="{
            color: !errorMessage ? (isFocus ? focusColor : blurColor) : undefined,
          }"
          :for="id"
        >
          {{ placeholder }}
        </label>
      </div>

      <div class="var-input__icon" :class="[!hint ? 'var-input--non-hint' : null]">
        <slot name="append-icon">
          <var-icon
            class="var-input__clear-icon"
            var-input-cover
            name="close-circle"
            size="14px"
            v-if="clearable && !isEmpty(value)"
            @click="handleClear"
          />
        </slot>
      </div>
    </div>

    <div
      class="var-input__line"
      :class="[
        formDisabled || disabled ? 'var-input--line-disabled' : null,
        errorMessage ? 'var-input--line-error' : null,
      ]"
      :style="{ background: !errorMessage ? blurColor : undefined }"
      v-if="line"
    >
      <div
        class="var-input__dot"
        :class="[
          isFocus ? 'var-input--spread' : null,
          formDisabled || disabled ? 'var-input--line-disabled' : null,
          errorMessage ? 'var-input--line-error' : null,
        ]"
        :style="{ background: !errorMessage ? focusColor : undefined }"
      ></div>
    </div>

    <var-form-details :error-message="errorMessage" :maxlength-text="maxlengthText" />
  </div>
</template>

<script>
import VarFormDetails from '../form-details'
import VarIcon from '../icon'
import { defineComponent } from '../utils/create'
import { props } from './props'
import { isEmpty } from '../utils/shared'
import { ValidationMixin } from '../utils/mixins/validation'
import { createChildrenMixin } from '../utils/mixins/relation'

export default defineComponent({
  name: 'VarInput',

  mixins: [ValidationMixin, createChildrenMixin('form', { parentKey: 'form', childrenKey: 'formItems' })],

  components: {
    VarIcon,
    VarFormDetails,
  },

  props,

  data() {
    return {
      id: `var-input-${this._uid}`,
      isFocus: false,
    }
  },

  computed: {
    maxlengthText() {
      const { maxlength, value } = this

      if (!maxlength) return ''

      if (this.isEmpty(value)) return `0 / ${maxlength}`

      return `${String(value).length}/${maxlength}`
    },

    formReadonly() {
      return this.form?.readonly
    },

    formDisabled() {
      return this.form?.disabled
    },
  },

  methods: {
    // expose
    reset() {
      this.getListeners().onInput?.('')
      this.resetValidation()
    },

    // expose
    validate() {
      this._validate(this.rules, this.value)
    },

    // expose
    focus() {
      this.$refs.el.focus()
    },

    // expose
    blur() {
      this.$refs.el.blur()
    },

    isEmpty,

    validateWithTrigger(trigger) {
      this.$nextTick(() => {
        const { validateTrigger, rules, value } = this
        this._validateWithTrigger(validateTrigger, trigger, rules, value)
      })
    },

    computePlaceholderState() {
      const { hint, value } = this

      if (!hint && !this.isEmpty(value)) {
        return 'var-input--placeholder-hidden'
      }
      if (hint && (!this.isEmpty(value) || this.isFocus)) {
        return 'var-input--placeholder-hint'
      }
    },

    handleFocus(e) {
      this.isFocus = true

      this.getListeners().onFocus?.(e)
      this.validateWithTrigger('onFocus')
    },

    handleBlur(e) {
      this.isFocus = false

      this.getListeners().onBlur?.(e)
      this.validateWithTrigger('onBlur')
    },

    handleInput(e) {
      const { value } = e.target

      this.getListeners().onInput?.(value, e)
      this.validateWithTrigger('onInput')
    },

    handleChange(e) {
      const { value } = e.target

      this.getListeners().onChange?.(value, e)
      this.validateWithTrigger('onChange')
    },

    handleClear() {
      const { disabled, readonly, clearable, getListeners } = this
      const { onInput, onClear } = getListeners()

      if (this.form?.disabled || this.form?.readonly || disabled || readonly || !clearable) {
        return
      }

      onInput?.('')
      onClear?.('')

      this.validateWithTrigger('onClear')
    },

    handleClick(e) {
      const { disabled, getListeners } = this

      if (this.form?.disabled || disabled) {
        return
      }

      getListeners().onClick?.(e)

      this.validateWithTrigger('onClick')
    },
  },
})
</script>

<style lang="less">
@import '../styles/common';
@import '../icon/icon';
@import '../form-details/formDetails';
@import './input';
</style>
