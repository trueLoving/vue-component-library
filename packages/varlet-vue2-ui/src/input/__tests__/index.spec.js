import example from '../example'
import Input from '..'
import VarInput from '../Input'
import { mount } from '@vue/test-utils'
import Vue from 'vue'
import { delay } from '../../utils/jest'

test('test input example', () => {
  const wrapper = mount(example)
  expect(wrapper.html()).toMatchSnapshot()
  wrapper.destroy()
})

test('test input plugin', () => {
  Vue.use(Input)
  expect(Vue.component(Input.name)).toBeTruthy()
})

test('test input focus & blur', async () => {
  const onFocus = jest.fn()
  const onBlur = jest.fn()

  const wrapper = mount(VarInput, {
    listeners: {
      focus: onFocus,
      blur: onBlur,
    },
  })

  wrapper.vm.focus()
  await wrapper.find('.var-input__input').trigger('focus')
  expect(wrapper.html()).toMatchSnapshot()

  wrapper.vm.blur()
  await wrapper.find('.var-input__input').trigger('blur')
  expect(wrapper.html()).toMatchSnapshot()

  wrapper.destroy()
})

test('test input onInput & onChange & onClick', async () => {
  const onInput = jest.fn((value) => wrapper.setProps({ value }))
  const onChange = jest.fn()
  const onClick = jest.fn()

  const wrapper = mount(VarInput, {
    propsData: {
      value: '',
    },
    listeners: {
      input: onInput,
      change: onChange,
      click: onClick,
    },
  })

  await wrapper.trigger('click')
  expect(onClick).toHaveBeenCalledTimes(1)

  wrapper.find('.var-input__input').setValue('t')

  await wrapper.find('.var-input__input').trigger('input')
  expect(onInput).lastCalledWith('t', new Event('input'))
  expect(wrapper.props('value')).toBe('t')

  await wrapper.find('.var-input__input').trigger('change')
  expect(onChange).lastCalledWith('t', new Event('input'))

  wrapper.destroy()
})

test('test input maxlength', () => {
  const wrapper = mount(VarInput, {
    propsData: {
      value: 'text',
      maxlength: 100,
    },
  })

  expect(wrapper.find('.var-form-details__length').text()).toBe('4/100')
  expect(wrapper.html()).toMatchSnapshot()

  wrapper.destroy()
})

test('test input hint to be false', () => {
  const wrapper = mount(VarInput, {
    propsData: {
      value: 'text',
      hint: false,
    },
  })

  expect(wrapper.html()).toMatchSnapshot()

  wrapper.destroy()
})

test('test input clear', async () => {
  const onInput = jest.fn((value) => wrapper.setProps({ value }))
  const onClear = jest.fn()

  const wrapper = mount(VarInput, {
    propsData: {
      value: 'text',
      clearable: true,
    },
    listeners: {
      clear: onClear,
      input: onInput,
    },
  })

  expect(wrapper.html()).toMatchSnapshot()

  await wrapper.find('.var-input__clear-icon').trigger('click')
  expect(onClear).lastCalledWith('')
  expect(wrapper.props('value')).toBe('')

  wrapper.destroy()
})

const triggerEvents = async (wrapper) => {
  await wrapper.find('.var-input__input').trigger('input')
  await wrapper.find('.var-input__input').trigger('change')
  await wrapper.find('.var-input__clear-icon').trigger('click')
  await wrapper.trigger('click')
}

test('test input disabled', async () => {
  const onClear = jest.fn()
  const onClick = jest.fn()
  const onInput = jest.fn()
  const onChange = jest.fn()

  const wrapper = mount(VarInput, {
    propsData: {
      value: 'hello',
      clearable: true,
      disabled: true,
    },
    listeners: {
      input: onInput,
      clear: onClear,
      click: onClick,
      change: onChange,
    },
  })

  await triggerEvents(wrapper)

  expect(onInput).toHaveBeenCalledTimes(0)
  expect(onClear).toHaveBeenCalledTimes(0)
  expect(onClick).toHaveBeenCalledTimes(0)
  expect(onChange).toHaveBeenCalledTimes(0)

  wrapper.destroy()
})

test('test input readonly', async () => {
  const onClear = jest.fn()
  const onClick = jest.fn()
  const onInput = jest.fn()
  const onChange = jest.fn()

  const wrapper = mount(VarInput, {
    propsData: {
      value: 'hello',
      clearable: true,
      readonly: true,
    },
    listeners: {
      input: onInput,
      clear: onClear,
      click: onClick,
      chnage: onChange,
    },
  })

  await triggerEvents(wrapper)

  expect(onInput).toHaveBeenCalledTimes(0)
  expect(onClear).toHaveBeenCalledTimes(0)
  expect(onClick).toHaveBeenCalledTimes(2)
  expect(onChange).toHaveBeenCalledTimes(0)

  wrapper.destroy()
})

test('test input validation', async () => {
  const onInput = jest.fn((value) => wrapper.setProps({ value }))

  const wrapper = mount(VarInput, {
    propsData: {
      value: '',
      rules: [(v) => v.length > 3 || '长度必须大于3'],
    },
    listeners: {
      input: onInput,
    },
  })

  wrapper.find('.var-input__input').setValue('1')
  await wrapper.find('.var-input__input').trigger('input')
  await delay(16)
  expect(wrapper.find('.var-form-details__message').text()).toBe('长度必须大于3')
  expect(wrapper.html()).toMatchSnapshot()

  wrapper.vm.reset()
  await delay(16)
  expect(wrapper.props('value')).toBe('')
  expect(wrapper.html()).toMatchSnapshot()

  wrapper.find('.var-input__input').setValue('1234')
  await wrapper.find('.var-input__input').trigger('input')
  await delay(16)
  expect(wrapper.find('.var-form-details__message').exists()).toBeFalsy()
  expect(wrapper.html()).toMatchSnapshot()

  wrapper.destroy()
})
