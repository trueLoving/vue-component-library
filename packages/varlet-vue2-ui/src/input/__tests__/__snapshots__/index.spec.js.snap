// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test input clear 1`] = `
"<div class=\\"var-input var--box\\">
  <div class=\\"var-input__controller\\">
    <div class=\\"var-input__icon\\"></div>
    <div class=\\"var-input__wrap\\">
      <!----> <input autocomplete=\\"new-password\\" id=\\"var-input-58\\" type=\\"text\\" class=\\"var-input__input\\"> <label for=\\"var-input-58\\" class=\\"var--ellipsis var-input__placeholder var-input--placeholder-hint\\">

      </label></div>
    <div class=\\"var-input__icon\\"><i class=\\"var-icon var-input__clear-icon var-icon--set var-icon-close-circle\\" style=\\"transition: transform 0ms; font-size: 14px;\\" var-input-cover=\\"\\"></i></div>
  </div>
  <div class=\\"var-input__line\\">
    <div class=\\"var-input__dot\\"></div>
  </div>
  <transition-stub name=\\"var-form-details\\">
    <!---->
  </transition-stub>
</div>"
`;

exports[`test input example 1`] = `
"<div class=\\"var-input-example\\">
  <div class=\\"app-type pb\\">基本使用</div>
  <div class=\\"var-input var--box\\">
    <div class=\\"var-input__controller\\">
      <div class=\\"var-input__icon\\"></div>
      <div class=\\"var-input__wrap\\">
        <!----> <input autocomplete=\\"new-password\\" id=\\"var-input-3\\" type=\\"text\\" class=\\"var-input__input\\"> <label for=\\"var-input-3\\" class=\\"var--ellipsis var-input__placeholder\\">
          请输入文本
        </label></div>
      <div class=\\"var-input__icon\\">
        <!---->
      </div>
    </div>
    <div class=\\"var-input__line\\">
      <div class=\\"var-input__dot\\"></div>
    </div>
    <transition-stub name=\\"var-form-details\\">
      <!---->
    </transition-stub>
  </div>
  <div class=\\"app-type pb\\">朴素模式</div>
  <div class=\\"var-input var--box\\">
    <div class=\\"var-input__controller\\">
      <div class=\\"var-input__icon var-input--non-hint\\"></div>
      <div class=\\"var-input__wrap var-input--non-hint\\">
        <!----> <input autocomplete=\\"new-password\\" id=\\"var-input-7\\" type=\\"text\\" class=\\"var-input__input\\"> <label for=\\"var-input-7\\" class=\\"var--ellipsis var-input__placeholder var-input--placeholder-non-hint\\">
          请输入文本
        </label></div>
      <div class=\\"var-input__icon var-input--non-hint\\">
        <!---->
      </div>
    </div>
    <!---->
    <transition-stub name=\\"var-form-details\\">
      <!---->
    </transition-stub>
  </div>
  <div class=\\"app-type pb\\">文本域</div>
  <div class=\\"var-input var--box\\">
    <div class=\\"var-input__controller\\">
      <div class=\\"var-input__icon\\"></div>
      <div class=\\"var-input__wrap\\">
        <!----> <textarea autocomplete=\\"new-password\\" id=\\"var-input-11\\" type=\\"text\\" rows=\\"8\\" class=\\"var-input__input var-input--textarea\\" style=\\"resize: none;\\"></textarea> <label for=\\"var-input-11\\" class=\\"var--ellipsis var-input__textarea-placeholder\\">
          请输入文本
        </label></div>
      <div class=\\"var-input__icon\\">
        <!---->
      </div>
    </div>
    <div class=\\"var-input__line\\">
      <div class=\\"var-input__dot\\"></div>
    </div>
    <transition-stub name=\\"var-form-details\\">
      <!---->
    </transition-stub>
  </div>
  <div class=\\"app-type pb\\">最大长度</div>
  <div class=\\"var-input var--box\\">
    <div class=\\"var-input__controller\\">
      <div class=\\"var-input__icon\\"></div>
      <div class=\\"var-input__wrap\\">
        <!----> <input autocomplete=\\"new-password\\" id=\\"var-input-15\\" type=\\"text\\" maxlength=\\"10\\" class=\\"var-input__input\\"> <label for=\\"var-input-15\\" class=\\"var--ellipsis var-input__placeholder\\">
          请输入文本
        </label></div>
      <div class=\\"var-input__icon\\">
        <!---->
      </div>
    </div>
    <div class=\\"var-input__line\\">
      <div class=\\"var-input__dot\\"></div>
    </div>
    <transition-stub name=\\"var-form-details\\">
      <div class=\\"var-form-details\\">
        <div class=\\"var-form-details__message\\"></div>
        <div class=\\"var-form-details__length\\">0 / 10</div>
      </div>
    </transition-stub>
  </div>
  <div class=\\"app-type pb\\">禁用</div>
  <div class=\\"var-input var--box var-input--disabled\\">
    <div class=\\"var-input__controller var-input--disabled\\">
      <div class=\\"var-input__icon\\"></div>
      <div class=\\"var-input__wrap\\">
        <!----> <input autocomplete=\\"new-password\\" id=\\"var-input-19\\" disabled=\\"disabled\\" type=\\"text\\" class=\\"var-input__input var-input--disabled\\"> <label for=\\"var-input-19\\" class=\\"var--ellipsis var-input__placeholder\\">
          请输入文本
        </label></div>
      <div class=\\"var-input__icon\\">
        <!---->
      </div>
    </div>
    <div class=\\"var-input__line var-input--line-disabled\\">
      <div class=\\"var-input__dot var-input--line-disabled\\"></div>
    </div>
    <transition-stub name=\\"var-form-details\\">
      <!---->
    </transition-stub>
  </div>
  <div class=\\"app-type pb\\">只读</div>
  <div class=\\"var-input var--box\\">
    <div class=\\"var-input__controller\\">
      <div class=\\"var-input__icon\\"></div>
      <div class=\\"var-input__wrap\\">
        <!----> <input autocomplete=\\"new-password\\" id=\\"var-input-23\\" disabled=\\"disabled\\" type=\\"text\\" class=\\"var-input__input\\"> <label for=\\"var-input-23\\" class=\\"var--ellipsis var-input__placeholder\\">
          请输入文本
        </label></div>
      <div class=\\"var-input__icon\\">
        <!---->
      </div>
    </div>
    <div class=\\"var-input__line\\">
      <div class=\\"var-input__dot\\"></div>
    </div>
    <transition-stub name=\\"var-form-details\\">
      <!---->
    </transition-stub>
  </div>
  <div class=\\"app-type pb\\">可清除</div>
  <div class=\\"var-input var--box\\">
    <div class=\\"var-input__controller\\">
      <div class=\\"var-input__icon\\"></div>
      <div class=\\"var-input__wrap\\">
        <!----> <input autocomplete=\\"new-password\\" id=\\"var-input-27\\" type=\\"text\\" class=\\"var-input__input\\"> <label for=\\"var-input-27\\" class=\\"var--ellipsis var-input__placeholder var-input--placeholder-hint\\">
          请输入文本
        </label></div>
      <div class=\\"var-input__icon\\"><i class=\\"var-icon var-input__clear-icon var-icon--set var-icon-close-circle\\" style=\\"transition: transform 0ms; font-size: 14px;\\" var-input-cover=\\"\\"></i></div>
    </div>
    <div class=\\"var-input__line\\">
      <div class=\\"var-input__dot\\"></div>
    </div>
    <transition-stub name=\\"var-form-details\\">
      <!---->
    </transition-stub>
  </div>
  <div class=\\"app-type pb\\">显示图标</div>
  <div class=\\"var-input var--box\\">
    <div class=\\"var-input__controller\\">
      <div class=\\"var-input__icon\\"><i class=\\"var-icon prepend-icon var-icon--set var-icon-plus\\" style=\\"transition: transform 0ms;\\"></i></div>
      <div class=\\"var-input__wrap\\">
        <!----> <input autocomplete=\\"new-password\\" id=\\"var-input-32\\" type=\\"text\\" class=\\"var-input__input\\"> <label for=\\"var-input-32\\" class=\\"var--ellipsis var-input__placeholder\\">
          请输入文本
        </label></div>
      <div class=\\"var-input__icon\\"><i class=\\"var-icon append-icon var-icon--set var-icon-minus\\" style=\\"transition: transform 0ms;\\"></i></div>
    </div>
    <div class=\\"var-input__line\\">
      <div class=\\"var-input__dot\\"></div>
    </div>
    <transition-stub name=\\"var-form-details\\">
      <!---->
    </transition-stub>
  </div>
  <div class=\\"app-type\\">字段校验</div>
  <div class=\\"var-input var--box\\">
    <div class=\\"var-input__controller\\">
      <div class=\\"var-input__icon\\"></div>
      <div class=\\"var-input__wrap\\">
        <!----> <input autocomplete=\\"new-password\\" id=\\"var-input-38\\" type=\\"text\\" class=\\"var-input__input\\"> <label for=\\"var-input-38\\" class=\\"var--ellipsis var-input__placeholder\\">
          请输入文本
        </label></div>
      <div class=\\"var-input__icon\\">
        <!---->
      </div>
    </div>
    <div class=\\"var-input__line\\">
      <div class=\\"var-input__dot\\"></div>
    </div>
    <transition-stub name=\\"var-form-details\\">
      <!---->
    </transition-stub>
  </div>
  <div style=\\"height: 40px;\\"></div>
</div>"
`;

exports[`test input focus & blur 1`] = `
"<div class=\\"var-input var--box\\">
  <div class=\\"var-input__controller\\">
    <div class=\\"var-input__icon\\"></div>
    <div class=\\"var-input__wrap\\">
      <!----> <input autocomplete=\\"new-password\\" id=\\"var-input-42\\" type=\\"text\\" class=\\"var-input__input\\"> <label for=\\"var-input-42\\" class=\\"var--ellipsis var-input__placeholder\\">

      </label></div>
    <div class=\\"var-input__icon\\">
      <!---->
    </div>
  </div>
  <div class=\\"var-input__line\\">
    <div class=\\"var-input__dot\\"></div>
  </div>
  <transition-stub name=\\"var-form-details\\">
    <!---->
  </transition-stub>
</div>"
`;

exports[`test input focus & blur 2`] = `
"<div class=\\"var-input var--box\\">
  <div class=\\"var-input__controller\\">
    <div class=\\"var-input__icon\\"></div>
    <div class=\\"var-input__wrap\\">
      <!----> <input autocomplete=\\"new-password\\" id=\\"var-input-42\\" type=\\"text\\" class=\\"var-input__input\\"> <label for=\\"var-input-42\\" class=\\"var--ellipsis var-input__placeholder\\">

      </label></div>
    <div class=\\"var-input__icon\\">
      <!---->
    </div>
  </div>
  <div class=\\"var-input__line\\">
    <div class=\\"var-input__dot\\"></div>
  </div>
  <transition-stub name=\\"var-form-details\\">
    <!---->
  </transition-stub>
</div>"
`;

exports[`test input hint to be false 1`] = `
"<div class=\\"var-input var--box\\">
  <div class=\\"var-input__controller\\">
    <div class=\\"var-input__icon var-input--non-hint\\"></div>
    <div class=\\"var-input__wrap var-input--non-hint\\">
      <!----> <input autocomplete=\\"new-password\\" id=\\"var-input-54\\" type=\\"text\\" class=\\"var-input__input\\"> <label for=\\"var-input-54\\" class=\\"var--ellipsis var-input__placeholder var-input--placeholder-hidden var-input--placeholder-non-hint\\">

      </label></div>
    <div class=\\"var-input__icon var-input--non-hint\\">
      <!---->
    </div>
  </div>
  <div class=\\"var-input__line\\">
    <div class=\\"var-input__dot\\"></div>
  </div>
  <transition-stub name=\\"var-form-details\\">
    <!---->
  </transition-stub>
</div>"
`;

exports[`test input maxlength 1`] = `
"<div class=\\"var-input var--box\\">
  <div class=\\"var-input__controller\\">
    <div class=\\"var-input__icon\\"></div>
    <div class=\\"var-input__wrap\\">
      <!----> <input autocomplete=\\"new-password\\" id=\\"var-input-50\\" type=\\"text\\" maxlength=\\"100\\" class=\\"var-input__input\\"> <label for=\\"var-input-50\\" class=\\"var--ellipsis var-input__placeholder var-input--placeholder-hint\\">

      </label></div>
    <div class=\\"var-input__icon\\">
      <!---->
    </div>
  </div>
  <div class=\\"var-input__line\\">
    <div class=\\"var-input__dot\\"></div>
  </div>
  <transition-stub name=\\"var-form-details\\">
    <div class=\\"var-form-details\\">
      <div class=\\"var-form-details__message\\"></div>
      <div class=\\"var-form-details__length\\">4/100</div>
    </div>
  </transition-stub>
</div>"
`;

exports[`test input validation 1`] = `
"<div class=\\"var-input var--box\\">
  <div class=\\"var-input__controller var-input--error\\">
    <div class=\\"var-input__icon\\"></div>
    <div class=\\"var-input__wrap\\">
      <!----> <input autocomplete=\\"new-password\\" id=\\"var-input-73\\" type=\\"text\\" class=\\"var-input__input var-input--caret-error\\"> <label for=\\"var-input-73\\" class=\\"var--ellipsis var-input--error var-input__placeholder var-input--placeholder-hint\\">

      </label></div>
    <div class=\\"var-input__icon\\">
      <!---->
    </div>
  </div>
  <div class=\\"var-input__line var-input--line-error\\">
    <div class=\\"var-input__dot var-input--line-error\\"></div>
  </div>
  <transition-stub name=\\"var-form-details\\">
    <div class=\\"var-form-details\\">
      <div class=\\"var-form-details__message\\">长度必须大于3</div>
      <div class=\\"var-form-details__length\\"></div>
    </div>
  </transition-stub>
</div>"
`;

exports[`test input validation 2`] = `
"<div class=\\"var-input var--box\\">
  <div class=\\"var-input__controller\\">
    <div class=\\"var-input__icon\\"></div>
    <div class=\\"var-input__wrap\\">
      <!----> <input autocomplete=\\"new-password\\" id=\\"var-input-73\\" type=\\"text\\" class=\\"var-input__input\\"> <label for=\\"var-input-73\\" class=\\"var--ellipsis var-input__placeholder\\">

      </label></div>
    <div class=\\"var-input__icon\\">
      <!---->
    </div>
  </div>
  <div class=\\"var-input__line\\">
    <div class=\\"var-input__dot\\"></div>
  </div>
  <transition-stub name=\\"var-form-details\\">
    <!---->
  </transition-stub>
</div>"
`;

exports[`test input validation 3`] = `
"<div class=\\"var-input var--box\\">
  <div class=\\"var-input__controller\\">
    <div class=\\"var-input__icon\\"></div>
    <div class=\\"var-input__wrap\\">
      <!----> <input autocomplete=\\"new-password\\" id=\\"var-input-73\\" type=\\"text\\" class=\\"var-input__input\\"> <label for=\\"var-input-73\\" class=\\"var--ellipsis var-input__placeholder var-input--placeholder-hint\\">

      </label></div>
    <div class=\\"var-input__icon\\">
      <!---->
    </div>
  </div>
  <div class=\\"var-input__line\\">
    <div class=\\"var-input__dot\\"></div>
  </div>
  <transition-stub name=\\"var-form-details\\">
    <!---->
  </transition-stub>
</div>"
`;
